### ISSUE TITLE
Is there a way to get intellisense for the `fields` parameter in the can/cannot functions?

### ISSUE THREAD

**Issue by cerino-ligutom** [stalniy/casl#308]
I use TS and it would be nice to get intellisense for it.

`fields` parameter:
![image](https://user-images.githubusercontent.com/6721822/81457434-ad960c00-91c8-11ea-8056-75e2a59ab8e7.png)

Was thinking of something like maybe when we pass a class, it derives properties from the class as possible values OR we could explicitly set a generic for the can/cannot functions.

**Comment by stalniy** [comment]
Hi, 

Thanks for the issue! 
Initially, I also thought about it :) there are 2 reasons why I decided to type it as a `string`:

1. CASL supports wildcards in field names (https://stalniy.github.io/casl/v4/en/guide/restricting-fields#field-patterns)
2. CASL supports permissions on nested fields (e.g., `address.street`)

I guess, in both  these cases, it’s impossible to infer types for fields from classes. 

So, if you can know a good way to overcome that, please share your thoughts. 

Also, it’s important to note that different subjects may have different fields, so the types should be subject dependent (e.g., BlogPost has title and author but User has firstName and lastName).

**Comment by cerino-ligutom** [comment]
Ah... Maybe a union such that if the 2nd parameter is not of type string, then it starts deriving possible values based on the properties so the signature would be something like `keyof T | (keyof T)[] | string | string[] | undefined`. That'd work for the direct properties of the object. As for nested, that'd be a limitation as the `.` for nested properties cannot be inferred.

> Also, it’s important to note that different subjects may have different fields, so the types should be subject dependent (e.g., BlogPost has title and author but User has firstName and lastName).

Yes, that's why I was thinking the can/cannot functions could probably accept a generic (e.g. `can<T>`) so that it can derive properties from it.

Though I think, ultimately, an object (or array?) config would probably be the most type-safe (albeit verbose and quite messy to read 😢). I'll just throw random ideas below:

```ts
ability.can('read', 'User', 'address')

// Would be

// Accepts a string but the generic slot is filled
ability.can<User>('read', 'User', { address: true });
// Derive from the instantiated class
ability.can('read', instantiatedUserClass, { address: true });
// With the subject helper
ability.can('read', subject<User>('User', plainUserObject), { address: true })

```

As for nested ones, it's really messy though.
```ts
ability.can('read', 'User', 'address.city.location.name')

// to

ability.can('read', instantiatedUserClass, { 
  address: {
    city: {
      location: {
        name: true,
      }
    }
  }
});
```

Wildcards however would be tough if the pattern does not terminate with it.
```ts
ability.can('read', 'User', 'address.*.name')

ability.can('read', instantiatedUserClass, {
  address: {
    $wildcard: '*', // would need some metadata, either way this is ugly
    name: true,
  }
}
```

Let's look into an array config

```ts
ability.can('read', 'User', 'address')

// Would be

// Accepts a string but the generic slot is filled
ability.can<User>('read', 'User', ['address']);
// Derive from the instantiated class
ability.can('read', instantiatedUserClass, ['address']);
// With the subject helper
ability.can('read', subject<User>('User', plainUserObject), ['address'])
```

As for nested ones, a nested array would indicate "hierarchy" (dot notation).
```ts
ability.can('read', 'User', 'address.city.location.name')

// to

ability.can('read', instantiatedUserClass, [ ['address', ['city', ['location', ['name'] ] ] ] ]);
```

Wildcards could then be:
```ts
ability.can('read', 'User', 'address.*.name')

ability.can('read', instantiatedUserClass, ['address', ['*', ['name'] ] ])
```

I'm not sure how the type defs would look for the array one like but one thing for sure, a parser would be needed for both array/object configs.

**Comment by stalniy** [comment]
Yes, pretty hard to express a very simple concept - object field :) There are actually ways to support nested fields using `[“address”, “street”]` syntax but It will be a bit more complicated to express wildcards especially taking into consideration that this is opt-in feature. Types for the project even now quite complicated, and I don’t want to introduce additional complexity for fields. At least not right now

That’s why I decided to use weaker typing and used `string` type for field

**Comment by stalniy** [comment]
I’m going to close this because  I don’t plan to implement it in the nearest future. If there will be more requests to add support for field typing, I’ll reconsider my decision. 

Thanks for using CASL!

**Comment by stalniy** [comment]
This will be partially implemented in #333, we will get intellisense for fields but not strict restriction (only by passing 2nd generic parameter). 

Will be available in v5
